src/NHibernate/Connection/ResilientDriverConnectionProvider.cs

```diff
@@ -0,0 +1,97 @@
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.Common;
+using System.Threading;
+using NHibernate;
+using NHibernate.Connection;
+
+namespace NHibernate.Connection
+{
+	public class ResilientDriverConnectionProvider : DriverConnectionProvider
+	{
+		public const String ConnectionDelayBetweenTries = "connection.delay_between_tries";
+		public const String ConnectionMaxTries = "connection.max_tries";
+
+		private static readonly IInternalLogger log = LoggerProvider.LoggerFor(typeof(ResilientDriverConnectionProvider));
+
+		public ResilientDriverConnectionProvider()
+		{
+			this.MaxTries = 3;
+			this.DelayBetweenTries = TimeSpan.FromSeconds(5);
+		}
+
+		public Int32 MaxTries { get; set; }
+
+		public TimeSpan DelayBetweenTries { get; set; }
+
+		public override void Configure(IDictionary<String, String> settings)
+		{
+			String maxTries;
+			String delayBetweenTries;
+
+			if (settings.TryGetValue(ConnectionMaxTries, out maxTries) == true)
+			{
+				this.MaxTries = Int32.Parse(maxTries);
+			}
+
+			if (settings.TryGetValue(ConnectionDelayBetweenTries, out delayBetweenTries) == true)
+			{
+				TimeSpan timespan;
+				Int32 delay;
+
+				if (TimeSpan.TryParse(delayBetweenTries, out timespan) == true)
+				{
+					this.DelayBetweenTries = timespan;
+				}
+				else if (Int32.TryParse(delayBetweenTries, out delay) == true)
+				{
+					this.DelayBetweenTries = TimeSpan.FromSeconds(delay);
+				}
+				else
+				{
+					throw (new Exception("Invalid delay"));
+				}
+			}
+
+			base.Configure(settings);
+		}
+
+		public override IDbConnection GetConnection()
+		{
+			IDbConnection con = null;
+
+			for (var i = 0; i < this.MaxTries; ++i)
+			{
+				try
+				{
+					log.Debug(String.Format("Attempting to get connection, {0} of {1}", (i + 1), this.MaxTries));
+					con = base.GetConnection();
+
+					if (con.State != ConnectionState.Open)
+					{
+						con.Open();
+					}
+
+					log.Debug(String.Concat("Got a connection after {0} tries", (i + 1)));
+
+					break;
+				}
+				catch (Exception ex)
+				{
+					if (i == this.MaxTries - 1)
+					{
+						log.Error(String.Concat("Could not get connection after {0} tries", this.MaxTries), ex);
+						throw;
+					}
+					else
+					{
+						Thread.Sleep(this.DelayBetweenTries);
```

Comment:
In case of an error consider clearing the connection pools. Our server became unusable and required retarting after the SQL Server had been swapped by some MS load balancing or availability technology. The reason was that all connections in the pool got broken. Trying to get a new connection didn't help, because it came from the pool again. It probably won't hurt to clear the pools here. When no connection can be established, it's very unlikely that the pool has anything useful. Unfortunately, I've never got the chance to test whether clearing the pool really solves the problem.